WebGPURenderer: MSAA, Postprocessing and Wireframe support in the WebGL Backend

RenaudRohlinger opened the pull request:
Happy New Year!!

Multiple points are addressed in this PR:
- MSAA Support
- Wireframe Support
- Postprocessing Support in the WebGLBackend
- Removed WebGPU condition on `webgpu_postprocessing_afterimage`, `webgpu_skinning_instancing`, `webgpu_portal`.
- Fix RenderTarget broken after resize in WebGLBackend #27466 

Live example MSAA:
[webgpu_multisampled_renderbuffers](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_multisampled_renderbuffers.html)

Live example Postprocessing:
- [webgpu_postprocessing_afterimage](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_postprocessing_afterimage.html)
- [webgpu_skinning_instancing](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_skinning_instancing.html)
- [webgpu_portal](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_portal.html)

[![image](https://github.com/mrdoob/three.js/assets/15867665/e4ae3c12-b26d-4238-b2ea-b374ad5fab5f)](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_multisampled_renderbuffers.html)

_This contribution is funded by [Utsubo](https://utsubo.com/)_

RenaudRohlinger committed:
support material.wireframe in webgl backend

Mugen87 commented on examples/jsm/renderers/webgl/WebGLBackend.js:
```diff
@@ -386,7 +386,20 @@ class WebGLBackend extends Backend {
-		else mode = gl.TRIANGLES;
+		else {
+
+			if ( material.wireframe === true ) {
```
Don't we need a wireframe attribute similar to `WebGLRenderer`?

https://github.com/mrdoob/three.js/blob/69e8505887d7c0c847515bcea2114a4ab8040c36/src/renderers/WebGLRenderer.js#L795-L797

RenaudRohlinger replied:
Not sure about this one. If that's ok I would like to improve this part in another PR.

sunag replied:
`WebGPURenderer` handle it in `Geometries.js` and provide the attribute for both backends. We must use `renderObject.getIndex()` to obtain the corresponding indexes.

RenaudRohlinger committed:
fix reattach framebuffer, support msaa and wireframe demo

RenaudRohlinger commented:
I think it would be counter productive to try to split the development of WebGPURenderer into tons of PR in the current state.
Added to this PR some fixes notably #27466, and MSAA support and a new webgpu example that combine all this (`webgpu_multisampled_renderbuffers`).

RenaudRohlinger committed:
fix depth samples === 0

RenaudRohlinger committed:
cleanup

RenaudRohlinger committed:
more cleanup

RenaudRohlinger committed:
merge and cleaup

RenaudRohlinger committed:
more cleanup to the webgl pipeline

RenaudRohlinger commented:
Cleaned up a bit more the code and made some space and improvement for another issue I wanted to fix (postprocessing support in the WebGL backend).

I'm still struggling with a feedback loop issue from the WebGL Backend when multiple framebuffers are involved (which is currently an issue in the dev branch):
`WebGL warning: drawArraysInstanced: Texture level 0 would be read by TEXTURE_2D unit 0, but written by framebuffer attachment COLOR_ATTACHMENT0, which would be illegal feedback.`

Not sure how I should update the pipeline there and where to unbind the previous RTT which is then attached as a texture to next one. But maybe we could merge this PR before it grows even larger.

/cc @sunag

RenaudRohlinger committed:
postprocessing support

RenaudRohlinger commented:
Nevermind, also added support for postprocessing /cc @sunag 😄

RenaudRohlinger committed:
fix msaa

RenaudRohlinger committed:
fix example afterimage

RenaudRohlinger committed:
improved state management and performances

RenaudRohlinger committed:
generate a new webgl slot for fbos

RenaudRohlinger committed:
revert puppeteer

RenaudRohlinger committed:
fix copyFrameBufferToTexture with correct unbinding and cleanup

RenaudRohlinger committed:
fix multipass pipeline in webgl backend and enable new demos

RenaudRohlinger committed:
regenerate screenshot with gaussian blur

RenaudRohlinger commented:
Added support for multipass in the postprocessing pipeline and removed the WebGPU condition in 2 new demos:

Live example here:
- [webgpu_skinning_instancing](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_skinning_instancing.html)
- [webgpu_portal](https://raw.githack.com/renaudrohlinger/three.js/webgpu-wireframe/examples/webgpu_skinning_instancing.html)

I'm currently re-binding the framebuffer in the WebGLBackend per render object in the `draw` method of the `WebGLBackend`.
This solution is not final as it requires each quad in the RTT pipeline to be in the same configuration (size and so on), the issue is that currently all the renderTarget setup is configured at the very beginning of `Renderer.render()`, but then if there is multiple pass in the pipeline each FBO will have its own `renderContext`. Without re-binding the framebuffer they will all inherit only of the configuration of the first `renderContext` called at the very beginning of `render()` which is not good. /cc @sunag 

I think there could be a better solution which might require a refactor in the `Renderer` class which will impact the WebGPU side so I'd rather not touch it in this PR.

I'm satisfied of the current state of this. I will stop there and create new PRs once this one is merged.

sunag commented:
I think `beginRender()` and `finishRender` could dealing with the re-binding of framebuffer, perhaps related to `renderContextData.previousContext`, the previous context information may be lost if it is called in a more than once in the same flow of `Node.updateBefore` maybe it can be related with re-bind in the `draw`. But I'm also in favor of checking this more calmly in other PR.  Amazing progress you did these days!

aardgoose commented:
I'd agree that framebuffer binding should not be in draw().  The current WebGLRenderer maintains a stack of contexts, rather than a single 'previous context' which would be a better solution. Multiple passes already have their own nested calls to beginRender() and finishRender().

One performance issue I see with the nested render passes is the multiple revalidations of the cache key for the lighting environment which is per frameId which changes per pass.
